### PR TITLE
More NOD optimizations

### PR DESCRIPTION
**Pull request by anothermattbrown**
- Specialize the NeighborUnreachable conditions and ipsByHostname in SynthesizerInputImpl
- Move global constraints (that must be true for every flow) like sane, srcNatted, and the query's headerspace constraint out of NOD and into SMT.
- Strengthen specialization to detect more unusable rules (i.e. those with always-false constraints)
- Strengthen simplification rules

### CODE REVIEW

**Comment by dhalperi** [comment]
:lgtm: but probably Ari should take a look.

---

Reviewed 16 of 16 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[projects/batfish/src/main/java/org/batfish/z3/IpSpaceSpecializer.java, line 141 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1149#-L9pj2p06kxUbApDiBu0:-L9pj2p1CLxxLh6R982h:bmjqca0) ([raw file](https://github.com/batfish/batfish/blob/f513b6b39b5dd6cf3441d6eaca985835805c4809/projects/batfish/src/main/java/org/batfish/z3/IpSpaceSpecializer.java#L141)):*
> ```Java
>     /*
>      * Remove any whitelisted Ips that are either blacklisted or don't overlap with the specialized
>      * whitelist.
> ```

I can't grok these comments in a way that makes this a change.

If the specialized whitelist is only things that are not covered by the blacklist, and we filter things that aren't blacklisted, how is this an new behavior?

(Just perf?)

---

*[projects/batfish/src/main/java/org/batfish/z3/Z3ContextJob.java, line 182 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1149#-L9pksYRArBv7sAYci1U:-L9pksYRArBv7sAYci1V:b9u0wt5) ([raw file](https://github.com/batfish/batfish/blob/f513b6b39b5dd6cf3441d6eaca985835805c4809/projects/batfish/src/main/java/org/batfish/z3/Z3ContextJob.java#L182)):*
<details><summary><i alt="summary">Quoted 8 lines of code…</i></summary>

> ```
>     /*
>     System.out.println(answer);
>     CommonUtil.writeFile(
>         Paths.get(String.format("/tmp/answer-%s.smt2", Instant.now().toString())),
>         answer.toString());
>         */
>     //    CommonUtil.writeFile(_settings.getQuestionPath().resolve("answer.smt2"),
>     // answer.toString());
> ```
</details>


delete?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1149#-:-L9pllWc4HQ6RQcoJCz_:bd6bn9p)*

**Comment by arifogel** [comment]
Reviewed 16 of 16 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[projects/batfish/src/main/java/org/batfish/z3/NodContext.java, line 57 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1149#-L9pku9SD8vKkORNV_pN:-L9pku9SD8vKkORNV_pO:b-af69hz) ([raw file](https://github.com/batfish/batfish/blob/f513b6b39b5dd6cf3441d6eaca985835805c4809/projects/batfish/src/main/java/org/batfish/z3/NodContext.java#L57)):*
> ```Java
>                                     // dummy rule that includes the smt constraint
>                                     new BasicRuleStatement(
>                                         program.getSmtConstraint(), Debug.INSTANCE))
> ```

Should this go in production?

---

*[projects/batfish/src/main/java/org/batfish/z3/expr/HeaderSpaceMatchExpr.java, line 405 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1149#-L9pmWBW-mGKbqhvIUxi:-L9pmWBW-mGKbqhvIUxj:bgrs6rn) ([raw file](https://github.com/batfish/batfish/blob/f513b6b39b5dd6cf3441d6eaca985835805c4809/projects/batfish/src/main/java/org/batfish/z3/expr/HeaderSpaceMatchExpr.java#L405)):*
> ```Java
>   }
> 
>   public static BooleanExpr matchOrigSrcOrDstIp(Set<IpWildcard> ipWildcards) {
> ```

When would you use this? The semantics are not at all obvious from the name.

---

*[projects/batfish/src/test/java/org/batfish/z3/NodJobTest.java, line 258 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1149#-L9pn6sb-izhXVWnAbfD:-L9pn6sb-izhXVWnAbfE:bzdqv4j) ([raw file](https://github.com/batfish/batfish/blob/f513b6b39b5dd6cf3441d6eaca985835805c4809/projects/batfish/src/test/java/org/batfish/z3/NodJobTest.java#L258)):*
> ```Java
>     Map<OriginateVrf, Map<String, Long>> fieldConstraintsByOriginateVrf =
>         nodJob.getOriginateVrfConstraints(z3Context, smtInput);
>     assertThat(fieldConstraintsByOriginateVrf.entrySet(), hasSize(0));
> ```

I am having trouble understanding this. If there are no constraints, it's unsatisfiable? Wouldn't that mean it's trivially satisfiable?

---

*[projects/batfish/src/test/java/org/batfish/z3/NodJobTest.java, line 316 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1149#-L9pnP2u18XU33RCQh0f:-L9pnP2u18XU33RCQh0g:bdvcyhj) ([raw file](https://github.com/batfish/batfish/blob/f513b6b39b5dd6cf3441d6eaca985835805c4809/projects/batfish/src/test/java/org/batfish/z3/NodJobTest.java#L316)):*
> ```Java
>     Map<OriginateVrf, Map<String, Long>> fieldConstraintsByOriginateVrf =
>         nodJob.getOriginateVrfConstraints(z3Context, smtInput);
>     assertThat(fieldConstraintsByOriginateVrf.entrySet(), hasSize(greaterThan(0)));
> ```

Doesn't this just mean it's satisfiable but not trivially so?

---

*[projects/batfish/src/test/java/org/batfish/z3/NodJobTest.java, line 332 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1149#-L9pnTqg7QGlLkagZXrk:-L9pnTqg7QGlLkagZXrl:bzdqv4j) ([raw file](https://github.com/batfish/batfish/blob/f513b6b39b5dd6cf3441d6eaca985835805c4809/projects/batfish/src/test/java/org/batfish/z3/NodJobTest.java#L332)):*
> ```Java
>     Map<OriginateVrf, Map<String, Long>> fieldConstraintsByOriginateVrf =
>         nodJob.getOriginateVrfConstraints(z3Context, smtInput);
>     assertThat(fieldConstraintsByOriginateVrf.entrySet(), hasSize(0));
> ```

I am having trouble understanding this. If there are no constraints, it's unsatisfiable? Wouldn't that mean it's trivially satisfiable?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1149)*

**Comment by arifogel** [comment]
Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[projects/batfish/src/test/java/org/batfish/z3/NodJobTest.java, line 258 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1149#-L9pn6sb-izhXVWnAbfD:-L9qctnNC3f8ktIc0Bwd:b489lj9) ([raw file](https://github.com/batfish/batfish/blob/f513b6b39b5dd6cf3441d6eaca985835805c4809/projects/batfish/src/test/java/org/batfish/z3/NodJobTest.java#L258)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

There are constraints. Specifically srcIp=3.0.0.0 and srcNatted=false, which are UNSAT for the NodJob constructed by `getNodJob`.  

I'm using `computeSmtInput`/`getOriginateVrfConstraints` instead of `computeNodSat` because the latter only checks the output from NOD. Since I moved headerspace, sane, and srcNatted constraints out of NOD and into SMT, we need to include the SMT step here.
</blockquote></details>

Placeholder: please make changes discussed offline (fixes reviewable erroneous red convo)

---

*[projects/batfish/src/test/java/org/batfish/z3/NodJobTest.java, line 316 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1149#-L9pnP2u18XU33RCQh0f:-L9qd4is4-Q5ospo469w:b-dm0h3v) ([raw file](https://github.com/batfish/batfish/blob/f513b6b39b5dd6cf3441d6eaca985835805c4809/projects/batfish/src/test/java/org/batfish/z3/NodJobTest.java#L316)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

Yes these constraints (srcIp=3.0.0.1 and srcNatted=false) should be satisfiable.
</blockquote></details>

Placeholder: please make changes discussed offline (fixes reviewable erroneous red convo)

---

*[projects/batfish/src/test/java/org/batfish/z3/NodJobTest.java, line 332 at r1](https://beta.reviewable.io/reviews/batfish/batfish/1149#-L9pnTqg7QGlLkagZXrk:-L9qd56k03uA8tfiNYR7:b-dm0h3v) ([raw file](https://github.com/batfish/batfish/blob/f513b6b39b5dd6cf3441d6eaca985835805c4809/projects/batfish/src/test/java/org/batfish/z3/NodJobTest.java#L332)):*
<details><summary><i>Previously, anothermattbrown (Matt Brown) wrote…</i></summary><blockquote>

There are constraints (srcIp=3.0.0.1 and srcNatted=true), and they're unsatisfiable.
</blockquote></details>

Placeholder: please make changes discussed offline (fixes reviewable erroneous red convo)

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1149)*

**Comment by dhalperi** [comment]
Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1149)*

**Comment by arifogel** [comment]
:lgtm: once conflicts are fixed.

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1149#-:-L9qq0L95TcMf-qSpV0m:biqea51)*

**Comment by arifogel** [comment]
Reviewed 3 of 4 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/1149)*

File: projects/batfish/src/main/java/org/batfish/z3/Z3ContextJob.java [hunk 1]
```diff
@@ -180,20 +180,25 @@ protected BoolExpr getSolverInput(Expr answer, NodProgram program, boolean negat
...
+              .getVariableNames()
+              .stream()
+              .filter(
+                  name -> !TransformationHeaderField.transformationHeaderFieldNames.contains(name))
+              .map(variablesAsConsts::get)
+              .collect(Collectors.toList());
+      List<BitVecExpr> reversedVars = Lists.reverse(vars);
+
+      Expr substitutedSmtConstraint =
+          program.getSmtConstraint().substituteVars(vars.toArray(new Expr[] {}));
```
Comment (jkhourybbn) [inline projects/batfish/src/main/java/org/batfish/z3/Z3ContextJob.java]: @anothermattbrown  I think we can make this more generic/useful: 

if program does not have an `smtConstraint` i.e., it defaults to the `TrueExpr.Instance`, introducing this `substitutedSmtConstraint` adds unnecessary complexity to the answer

Some of the reachability analysis I do only cares about the raw fixed point answer

Now with the line above, I get
```
(and <raw_fixed_point_answer_expression> true true)
```
So I am hacking the 'solverInput` answer I get back and extracting `solverInput.getArgs()[0]` to get my raw answer

In summary, if there is no smtconstraint, perhaps we can avoid complicating the answer

Comment (anothermattbrown) [reply projects/batfish/src/main/java/org/batfish/z3/Z3ContextJob.java]: Makes sense to me.
